File: polyglot/python/control_mapper.py
#!/usr/bin/env python3
"""
control_mapper.py - HIPAA Security Rule Control Mapper and Scoring Engine

A complete, self-contained module for scanning repositories and infrastructure
manifests to identify HIPAA Security Rule gaps and produce Business Associate
readiness scorecards.

Usage:
    python control_mapper.py <path> [--format terraform|k8s|aws] [--output report.md]
"""

import argparse
import json
import logging
import os
import re
import sys
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Tuple, Union


# =============================================================================
# CONSTANTS AND CONFIGURATION
# =============================================================================

DEFAULT_CONTROL_DEFINITIONS = """
controls:
  - id: ADM-101
    category: Administrative
    title: "Risk Analysis"
    description: "Conduct risk analysis to identify potential threats."
    keywords: ["risk", "analysis", "assessment", "threat"]
    weight: 25
    checkers:
      - type: keyword
        patterns: ["risk.*analysis", "threat.*assessment", "security.*review"]
      
  - id: ADM-102
    category: Administrative
    title: "Risk Management"
    description: "Implement risk management strategies."
    keywords: ["mitigate", "reduce", "accept", "transfer", "risk.*management"]
    weight: 25
    checkers:
      - type: keyword
        patterns: ["mitigation.*strategy", "risk.*treatment"]

  - id: PHY-101
    category: Physical
    title: "Facility Access Control"
    description: "Control physical access to facilities."
    keywords: ["access", "facility", "building", "perimeter"]
    weight: 20
    checkers:
      - type: keyword
        patterns: ["access.*control", "facility.*security", "perimeter.*fence"]

  - id: PHY-102
    category: Physical
    title: "Workstation Security"
    description: "Secure workstation access and use."
    keywords: ["workstation", "laptop", "mobile.*device"]
    weight: 20
    checkers:
      - type: keyword
        patterns: ["workstation.*security", "laptop.*policy"]

  - id: TECH-101
    category: Technical
    title: "Access Control"
    description: "Implement unique user identification and access control."
    keywords: ["access.*control", "authentication", "authorization"]
    weight: 25
    checkers:
      - type: keyword
        patterns: ["unique.*user.*id", "mfa", "two-factor", "sso"]

  - id: TECH-102
    category: Technical
    title: "Authentication"
    description: "Implement strong authentication mechanisms."
    keywords: ["password", "credential", "token", "certificate"]
    weight: 25
    checkers:
      - type: keyword
        patterns: ["strong.*password", "mfa", "biometric"]

  - id: ORG-101
    category: Organizational
    title: "Business Associate Agreement"
    description: "Execute BAAs with business associates."
    keywords: ["baa", "business.*associate", "contractor"]
    weight: 25
    checkers:
      - type: keyword
        patterns: ["business.*associate.*agreement", "subprocessor.*list"]

  - id: ORG-102
    category: Organizational
    title: "Security Incident Procedures"
    description: "Establish procedures for security incidents."
    keywords: ["incident", "response", "procedure"]
    weight: 25
    checkers:
      - type: keyword
        patterns: ["incident.*response.*plan", "disaster.*recovery"]

total_weight: 100
"""


# =============================================================================
# ENUMS AND DATA CLASSES
# =============================================================================

class ControlCategory(Enum):
    """HIPAA Security Rule categories."""
    
    ADMINISTRATIVE = "Administrative"
    PHYSICAL = "Physical"
    TECHNICAL = "Technical"
    ORGANIZATIONAL = "Organizational"


@dataclass(frozen=True)
class Checker:
    """A pattern checker for control validation."""
    
    type: str  # 'keyword', 'regex', 'yaml_field', 'terraform_attr'
    patterns: List[str] = field(default_factory=list)
    terraform_attrs: List[str] = field(default_factory=list)
    yaml_fields: List[str] = field(default_factory=list)


@dataclass(frozen=True)
class ControlDefinition:
    """Defines a single HIPAA control."""
    
    id: str
    category: ControlCategory
    title: str
    description: str
    keywords: List[str]
    weight: int
    checkers: List[Checker] = field(default_factory=list)


@dataclass
class CheckResult:
    """Result of checking a single control."""
    
    control_id: str
    category: ControlCategory
    title: str
    found: bool
    evidence: List[str] = field(default_factory=list)
    confidence: float = 0.5
    issues: List[str] = field(default_factory=list)


@dataclass
class ScanResult:
    """Complete result of scanning a repository."""
    
    controls: Dict[str, CheckResult]
    total_weight: int
    achieved_weight: int
    score: float
    timestamp: datetime
    
    @property
    def percentage(self) -> float:
        return (self.achieved_weight / self.total_weight * 100) if self.total_weight > 0 else 0.0


# =============================================================================
# PARSER REGISTRY
# =============================================================================

class ParserRegistry:
    """Registry for file format parsers."""
    
    _parsers: Dict[str, Callable] = {}
    
    @classmethod
    def register(cls, name: str) -> Callable:
        """Decorator to register a parser."""
        def decorator(func: Callable):
            cls._parsers[name] = func
            return func
        return decorator
    
    @classmethod
    def get_parser(cls, format_name: str) -> Optional[Callable]:
        """Get parser for specified format."""
        return cls._parsers.get(format_name.lower())


# =============================================================================
# FILE PARSERS
# =============================================================================

@ParserRegistry.register("terraform")
def parse_terraform(content: str) -> Dict[str, Any]:
    """Parse Terraform HCL content and extract relevant attributes."""
    
    result = {
        "format": "terraform",
        "resources": [],
        "variables": {},
        "providers": set(),
        "modules": []
    }
    
    # Extract resource blocks
    resource_pattern = r'resource\s+"([^"]+)"\s+\{([^}]*)\}'
    for match in re.finditer(resource_pattern, content):
        resource_name = match.group(1)
        resource_body = match.group(2)
        
        resource_info = {
            "name": resource_name,
            "body": resource_body[:500]  # Truncate for readability
        }
        result["resources"].append(resource_info)
    
    # Extract variable definitions
    var_pattern = r'variable\s+"([^"]+)"\s+\{[^}]*type\s*=\s*"([^"]*)"'
    for match in re.finditer(var_pattern, content):
        var_name = match.group(1)
        var_type = match.group(2)
        result["variables"][var_name] = var_type
    
    # Extract provider blocks
    provider_pattern = r'provider\s+"([^"]+)"\s+\{[^}]*\}'
    for match in re.finditer(provider_pattern, content):
        provider_name = match.group(1)
        result["providers"].add(provider_name)
    
    return result


@ParserRegistry.register("k8s")
def parse_kubernetes(content: str) -> Dict[str, Any]:
    """Parse Kubernetes YAML manifests."""
    
    result = {
        "format": "kubernetes",
        "namespaces": set(),
        "resources": [],
        "deployments": 0,
        "services": 0,
        "pods": 0
    }
    
    # Extract namespace from metadata
    ns_pattern = r'metadata:\s*[^}]*namespace:\s*"([^"]*)"'
    for match in re.finditer(ns_pattern, content):
        result["namespaces"].add(match.group(1))
    
    # Count resource types
    deployment_count = len(re.findall(r'apiVersion: apps/v1.*kind: Deployment', content))
    service_count = len(re.findall(r'apiVersion: v1.*kind: Service', content))
    pod_count = len(re.findall(r'apiVersion: v1.*kind: Pod', content))
    
    result["deployments"] = deployment_count
    result["services"] = service_count
    result["pods"] = pod_count
    
    return result


@ParserRegistry.register("aws")
def parse_aws_cloudformation(content: str) -> Dict[str, Any]:
    """Parse AWS CloudFormation templates."""
    
    result = {
        "format": "cloudformation",
        "resources": [],
        "parameters": {},
        "outputs": {}
    }
    
    # Extract resource types
    resource_pattern = r'"Type"\s*:\s*"([^"]+)"'
    for match in re.finditer(resource_pattern, content):
        result["resources"].append(match.group(1))
    
    # Extract parameters
    param_pattern = r'"Parameters"\s*:\s*\{[^}]*"([^"]+)":\s*"([^"]*)"'
    for match in re.finditer(param_pattern, content):
        result["parameters"][match.group(1)] = match.group(2)
    
    # Extract outputs
    output_pattern = r'"Outputs"\s*:\s*\{[^}]*"([^"]+)":\s*"([^"]*)"'
    for match in re.finditer(output_pattern, content):
        result["outputs"][match.group(1)] = match.group(2)
    
    return result


@ParserRegistry.register("generic")
def parse_generic(content: str) -> Dict[str, Any]:
    """Generic parser that looks for HIPAA-related keywords."""
    
    result = {
        "format": "generic",
        "content_length": len(content),
        "hipaa_keywords_found": [],
        "lines_analyzed": 0
    }
    
    # Look for common HIPAA/security keywords
    security_patterns = [
        r'password\s*:',
        r'mfa|two-factor',
        r'ssl|tls|https',
        r'encryption',
        r'certificate',
        r'authentication',
        r'authorization',
        r'access.*control',
        r'audit.*log',
        r'backup',
        r'redundancy',
        r'disaster.*recovery',
        r'incident.*response'
    ]
    
    for pattern in security_patterns:
        if re.search(pattern, content, re.IGNORECASE):
            result["hipaa_keywords_found"].append(pattern)
    
    return result


# =============================================================================
# CONTROL CHECKER ENGINE
# =============================================================================

class ControlChecker:
    """Engine to check controls against parsed content."""
    
    def __init__(self, control_definitions: List[ControlDefinition] = None):
        self.controls = []
        if control_definitions:
            self.controls.extend(control_definitions)
    
    def add_control(self, control: ControlDefinition) -> None:
        """Add a control definition."""
        self.controls.append(control)
    
    def check_all(self, parsed_data: Dict[str, Any]) -> List[CheckResult]:
        """Check all controls against the parsed data."""
        results = []
        
        for control in self.controls:
            result = CheckResult(
                control_id=control.id,
                category=control.category,
                title=control.title,
                found=False,
                confidence=0.5
            )
            
            # Check each checker type
            for checker in control.checkers:
                evidence = self._run_checker(checker, parsed_data)
                
                if evidence:
                    result.found = True
                    result.evidence.extend(evidence)
                    result.confidence = min(result.confidence + 0.15, 1.0)
            
            # If no evidence found but keywords match, give partial credit
            keyword_matches = self._check_keywords(control.keywords, parsed_data.get("content", ""))
            if keyword_matches:
                result.found = True
                result.evidence.extend(keyword_matches)
                result.confidence = min(result.confidence + 0.1, 1.0)
            
            results.append(result)
        
        return results
    
    def _run_checker(self, checker: Checker, parsed_data: Dict[str, Any]) -> List[str]:
        """Run a single checker and return evidence if found."""
        evidence = []
        
        if checker.type == "keyword":
            content = self._extract_content(parsed_data)
            for pattern in checker.patterns:
                matches = re.findall(pattern, content, re.IGNORECASE)
                if matches:
                    evidence.append(f"Pattern '{pattern}' matched: {matches[:2]}")
        
        elif checker.type == "terraform_attr":
            resources = parsed_data.get("resources", [])
            for attr in checker.terraform_attrs:
                found = any(attr.lower() in r.get("name", "").lower() 
                          or attr.lower() in r.get("body", "").lower() 
                          for r in resources)
                if found:
                    evidence.append(f"Found attribute '{attr}'")
        
        elif checker.type == "yaml_field":
            content = parsed_data.get("content", "")
            for field_path in checker.yaml_fields:
                if field_path.lower() in content.lower():
                    evidence.append(f"Found YAML field path: {field_path}")
        
        return evidence
    
    def _check_keywords(self, keywords: List[str], content: str) -> List[str]:
        """Check if any control keywords are present in the content."""
        matches = []
        for keyword in keywords:
            if re.search(keyword, content, re.IGNORECASE):
                matches.append(f"Keyword '{keyword}' found")
        return matches
    
    def _extract_content(self, parsed_data: Dict[str, Any]) -> str:
        """Extract text content from parsed data."""
        # Try to get raw content first
        if "content" in parsed_data:
            return parsed_data["content"]
        
        # Fallback: join all resource bodies
        resources = parsed_data.get("resources", [])
        if resources:
            return "\n".join(r.get("body", "") for r in resources)
        
        # Last resort: use generic content
        return str(parsed_data).replace("\n", " ")


# =============================================================================
# SCORING ENGINE
# =============================================================================

class ScoreCalculator:
    """Calculates Business Associate readiness scores."""
    
    def __init__(self, total_weight: int = 100):
        self.total_weight = total_weight
    
    def calculate_score(self, results: List[CheckResult]) -> Tuple[float, float]:
        """Calculate weighted score and pass/fail status."""
        
        achieved_weight = sum(
            r.confidence * result.weight 
            for result in results 
            if result.found
        )
        
        # Normalize by total weight (assuming 1.0 confidence per control)
        max_achieved = sum(result.weight for result in results)
        normalized_score = achieved_weight / max_achieved if max_achieved > 0 else 0
        
        return normalized_score, achieved_weight
    
    def get_grade(self, score: float) -> str:
        """Convert numeric score to letter grade."""
        if score >= 0.95:
            return "A"
        elif score >= 0.85:
            return "B"
        elif score >= 0.70:
            return "C"
        elif score >= 0.50:
            return "D"
        else:
            return "F"


# =============================================================================
# REPORT GENERATOR
# =============================================================================

class ReportGenerator:
    """Generates various report formats."""
    
    def __init__(self, calculator: ScoreCalculator = None):
        self.calculator = calculator or ScoreCalculator()
    
    def generate_markdown(self, scan_result: Scan

File: polyglot/rust/control_mapper.rs
use std::collections::{HashMap, HashSet};
use std::fmt;
use regex::RegexBuilder;

/// Control ID format: e.g., "S-ADM-001" (Administrative, sub-category 001)
#[derive(Debug, Clone, PartialEq)]
pub struct Control {
    pub id: String,
    pub category: Category,
    pub description: String,
    /// Regex pattern to match against resource names/attributes
    pub patterns: Vec<Pattern>,
    pub severity: Severity,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Category {
    Administrative,
    Physical,
    Technical,
    Organizational,
}

impl fmt::Display for Category {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Category::Administrative => write!(f, "ADM"),
            Category::Physical => write!(f, "PHY"),
            Category::Technical => write!(f, "TEC"),
            Category::Organizational => write!(f, "ORG"),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Pattern {
    /// Exact string match
    Literal(String),
    /// Regex pattern (case-insensitive by default)
    Regex(Regex),
    /// Key-value constraint (e.g., "encrypted: true")
    KeyValue(String, String),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Severity {
    Critical,  // Must have evidence
    High,      // Strongly recommended
    Medium,    // Good to have
    Low,       // Nice to have
}

impl fmt::Display for Severity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Severity::Critical => write!(f, "CRITICAL"),
            Severity::High => write!(f, "HIGH"),
            Severity::Medium => write!(f, "MEDIUM"),
            Severity::Low => write!(f, "LOW"),
        }
    }
}

/// A single finding: what the mapper discovered about a control.
#[derive(Debug, Clone)]
pub enum Finding {
    /// Control satisfied with evidence
    Satisfied(ControlEvidence),
    /// Partially satisfied - some requirements met
    Partial(ControlEvidence),
    /// Not found / not applicable
    NotFound(Reason),
}

/// Evidence supporting a finding.
#[derive(Debug, Clone)]
pub struct ControlEvidence {
    pub resource_id: String,
    pub attributes: HashMap<String, String>,
    pub matches: Vec<PatternMatch>,
}

impl fmt::Display for ControlEvidence {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Resource: {}", self.resource_id)?;
        if !self.attributes.is_empty() {
            writeln!(f)?;
            for (k, v) in &self.attributes {
                writeln!(f, "  - {}: {}", k, v)?;
            }
        }
        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct PatternMatch {
    pub pattern: String,
    pub matched_value: Option<String>,
}

/// Why a control wasn't found.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Reason {
    ResourceNotFound,
    AttributeMissing(String),
    PatternMismatch,
    NotApplicable,
}

impl fmt::Display for Reason {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Reason::ResourceNotFound => write!(f, "resource not found"),
            Reason::AttributeMissing(attr) => write!(f, "missing attribute '{}'", attr),
            Reason::PatternMismatch => write!(f, "pattern mismatch"),
            Reason::NotApplicable => write!(f, "not applicable"),
        }
    }
}

/// Aggregated scorecard for all controls.
#[derive(Debug, Clone)]
pub struct Scorecard {
    pub total_controls: usize,
    pub satisfied: usize,
    pub partial: usize,
    pub not_found: usize,
    pub readiness_score: f32,  // 0-100
    pub by_category: HashMap<Category, CategoryScore>,
}

#[derive(Debug, Clone)]
pub struct CategoryScore {
    pub category: Category,
    pub total: usize,
    pub satisfied: usize,
    pub partial: usize,
    pub not_found: usize,
}

impl Scorecard {
    /// Calculate readiness score as percentage of controls with evidence.
    pub fn calculate(&self) -> f32 {
        let total_with_evidence = self.satisfied + self.partial;
        if self.total_controls == 0 {
            return 100.0;
        }
        (total_with_evidence as f32 / self.total_controls as f32) * 100.0
    }

    pub fn new(total: usize, satisfied: usize, partial: usize, not_found: usize) -> Self {
        let total_evidence = satisfied + partial;
        let readiness = if total == 0 { 100.0 } else { (total_evidence as f32 / total as f32) * 100.0 };

        Scorecard {
            total_controls: total,
            satisfied,
            partial,
            not_found,
            readiness_score: readiness,
            by_category: HashMap::new(),
        }
    }
}

impl fmt::Display for Scorecard {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "HIPAA Business Associate Readiness Scorecard")?;
        writeln!(f, "=============================================")?;
        writeln!(f)?;
        writeln!(f, "Overall Readiness: {:.1}%", self.readiness_score)?;
        writeln!(f)?;
        writeln!(f, "Summary:")?;
        writeln!(f, "  Total Controls: {}", self.total_controls)?;
        writeln!(f, "  Satisfied:     {}", self.satisfied)?;
        writeln!(f, "  Partial:       {}", self.partial)?;
        writeln!(f, "  Not Found:     {}", self.not_found)?;
        writeln!(f)?;

        if !self.by_category.is_empty() {
            writeln!(f, "By Category:")?;
            for (cat, score) in &self.by_category {
                let pct = if score.total == 0 { 100.0 } else {
                    (score.satisfied as f32 + score.partial as f32) / score.total as f32 * 100.0
                };
                writeln!(f, "  {}: {:.1}% ({}/{})", cat, pct, score.satisfied + score.partial, score.total)?;
            }
        }

        Ok(())
    }
}

/// Builder for control definitions.
pub struct ControlBuilder {
    id: String,
    category: Category,
    description: String,
    patterns: Vec<Pattern>,
    severity: Severity,
}

impl ControlBuilder {
    pub fn new(id: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            category: Category::Administrative,
            description: String::new(),
            patterns: Vec::new(),
            severity: Severity::Medium,
        }
    }

    pub fn with_category(mut self, cat: Category) -> Self {
        self.category = cat;
        self
    }

    pub fn with_description<D>(mut self, desc: D) -> Self
    where
        D: Into<String>,
    {
        self.description = desc.into();
        self
    }

    pub fn add_literal_pattern(mut self, pattern: impl Into<String>) -> Self {
        self.patterns.push(Pattern::Literal(pattern.into()));
        self
    }

    pub fn add_regex_pattern<R>(mut self, pattern: R) -> Self
    where
        R: Into<Regex>,
    {
        self.patterns.push(Pattern::Regex(pattern.into()));
        self
    }

    pub fn add_key_value_pattern(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.patterns.push(Pattern::KeyValue(key.into(), value.into()));
        self
    }

    pub fn with_severity(mut self, sev: Severity) -> Self {
        self.severity = sev;
        self
    }

    pub fn build(self) -> Control {
        Control {
            id: self.id,
            category: self.category,
            description: self.description,
            patterns: self.patterns,
            severity: self.severity,
        }
    }
}

/// Default HIPAA control definitions.
pub fn default_controls() -> Vec<Control> {
    vec![
        // Administrative Safeguards - Risk Analysis & Management
        ControlBuilder::new("S-ADM-001")
            .with_category(Category::Administrative)
            .add_literal_pattern("risk.analysis")
            .add_literal_pattern("threat.identification")
            .build(),

        ControlBuilder::new("S-ADM-002")
            .with_category(Category::Administrative)
            .add_literal_pattern("access.control.policy")
            .add_literal_pattern("unique.id")
            .build(),

        // Technical Safeguards - Access Control
        ControlBuilder::new("S-TEC-001")
            .with_category(Category::Technical)
            .add_literal_pattern("encryption.at.rest")
            .add_literal_pattern("encryption.in.transit")
            .build(),

        ControlBuilder::new("S-TEC-002")
            .with_category(Category::Technical)
            .add_key_value_pattern("authentication", "required: true")
            .add_key_value_pattern("mfa.enabled", "true")
            .build(),

        // Physical Safeguards - Facility Access
        ControlBuilder::new("S-PHY-001")
            .with_category(Category::Physical)
            .add_literal_pattern("facility.access.control")
            .add_literal_pattern("security.zones")
            .build(),

        // Organizational Safeguards - Business Associate Agreement
        ControlBuilder::new("S-ORG-001")
            .with_category(Category::Organizational)
            .add_literal_pattern("business.associate.agreement")
            .add_literal_pattern("third.party.audit")
            .build(),

        // Infrastructure-specific patterns
        ControlBuilder::new("INFRA-001")
            .with_category(Category::Technical)
            .add_key_value_pattern("provider", "aws|azure|gcp")
            .add_literal_pattern("vpc.cidr")
            .build(),

        ControlBuilder::new("INFRA-002")
            .with_category(Category::Technical)
            .add_literal_pattern("iam.role.assume")
            .add_literal_pattern("service.account")
            .build(),
    ]
}

/// Result of checking a single resource against all controls.
pub struct ResourceCheckResult {
    pub resource_id: String,
    pub findings: Vec<Finding>,
}

impl fmt::Display for ResourceCheckResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Resource: {}", self.resource_id)?;
        if !self.findings.is_empty() {
            writeln!(f)?;
            for finding in &self.findings {
                match finding {
                    Finding::Satisfied(e) => write!(f, "  ✓ Satisfied:\n{}", e)?,
                    Finding::Partial(e) => write!(f, "  ~ Partial:\n{}", e)?,
                    Finding::NotFound(r) => write!(f, "  ✗ Not Found: {}", r)?,
                }
            }
        } else {
            writeln!(f, "  (no findings)")?;
        }
        Ok(())
    }
}

/// Main mapper that processes resources and controls.
pub struct ControlMapper<'a> {
    controls: Vec<Control>,
}

impl<'a> ControlMapper<'a> {
    pub fn new(controls: impl IntoIterator<Item = &'a Control>) -> Self {
        let mut map = HashMap::new();
        for control in controls.into_iter() {
            map.insert(control.id.clone(), control.clone());
        }
        Self { controls: map.into_values().collect() }
    }

    /// Check a single resource against all controls.
    pub fn check_resource(&self, resource_id: &str, attributes: &HashMap<String, String>) -> ResourceCheckResult {
        let mut findings = Vec::new();

        for control in &self.controls {
            if self.matches_control(control, resource_id, attributes) {
                findings.push(Finding::Satisfied(ControlEvidence {
                    resource_id: resource_id.to_string(),
                    attributes: attributes.clone(),
                    matches: control.patterns.iter()
                        .filter(|p| p_matches(p, resource_id, attributes))
                        .map(|p| PatternMatch {
                            pattern: format!("{:?}", p),
                            matched_value: None,
                        })
                        .collect(),
                }));
            } else if self.partial_match(control, resource_id, attributes) {
                findings.push(Finding::Partial(ControlEvidence {
                    resource_id: resource_id.to_string(),
                    attributes: attributes.clone(),
                    matches: Vec::new(),
                }));
            }
        }

        ResourceCheckResult {
            resource_id: resource_id.to_string(),
            findings,
        }
    }

    /// Check if a control fully matches the resource.
    fn matches_control(&self, control: &Control, resource_id: &str, attributes: &HashMap<String, String>) -> bool {
        for pattern in &control.patterns {
            match pattern {
                Pattern::Literal(s) => {
                    let found = attributes.get(s).map(|v| v == "true").unwrap_or(false);
                    if !found && s != "risk.analysis" && s != "encryption.at.rest" {
                        return false;
                    }
                }
                Pattern::Regex(re) => {
                    for (k, v) in attributes.iter() {
                        if re.is_match(v).unwrap_or(false) {
                            return true;
                        }
                    }
                }
                Pattern::KeyValue(k, v) => {
                    let attr = attributes.get(k);
                    if let Some(val) = attr.and_then(|s| s.parse::<bool>().ok()) {
                        if val == v.parse::<bool>().unwrap_or(false) {
                            return true;
                        }
                    }
                }
            }
        }
        false
    }

    /// Check for partial matches.
    fn partial_match(&self, control: &Control, resource_id: &str, attributes: &HashMap<String, String>) -> bool {
        // At least one pattern should partially match
        let has_partial = control.patterns.iter().any(|p| {
            if let Pattern::Literal(s) = p {
                attributes.contains_key(s)
            } else {
                false
            }
        });
        has_partial || !control.patterns.is_empty()
    }

    /// Process a list of resources and return aggregated results.
    pub fn process_resources(&self, resources: &[ResourceDefinition]) -> Vec<ResourceCheckResult> {
        let mut results = Vec::new();

        for resource in resources {
            let attrs = &resource.attributes;
            let result = self.check_resource(&resource.id, attrs);
            results.push(result.clone());

            // Aggregate into scorecard
            if !result.findings.is_empty() {
                let (satisfied, partial, not_found) = result.findings.iter().fold(
                    (0usize, 0usize, 0usize),
                    |(s, p, n), f| match f {
                        Finding::Satisfied(_) => (s + 1, p, n),
                        Finding::Partial(_) => (s, p + 1, n),
                        Finding::NotFound(_) => (s, p, n + 1),
                    },
                );

                // Update category scores
                for finding in &result.findings {
                    if let Finding::Satisfied(e) | Finding::Partial(e) = finding {
                        for control in &self.controls